### PR TITLE
[RFC] LibCore: UTF-8 agnostic Core::Path

### PR DESCRIPTION
**Pull request by DanShaders**
This PR introduces UTF-8 agnostic `Core::Path`, which internally stores paths as a pair of a directory file descriptor and 
a path relative to it. The relative part of a path is stored in a possibly zero-terminated byte buffer, so there is no need to copy the path just to add `\0` before making syscalls on non-Serenity platforms. The public API of `Core::Path` is somewhat constrained, so users are expected to rely on LibFileSystem for most of the tasks (like traversing directories, removing or creating them). This is done this way since it is notoriously difficult to do the aforementioned tasks without introducing TOCTOU errors in the case of concurrent access to FS.

Note that this is more an RFC rather than PR since it does not add any real uses of `Core::Path`. One can refer to tests to get an idea of how the class can be used. I have PoC tar implementation locally which does use `Core::Path` (and I can confirm that the exposed API is enough for it), however it requires substantial additional changes to `DirIterator`, `Directory`, `File` and `LibFileSystem`, so I decided to not include it here.

--------

(I made up the following a few days ago and found it really funny. Unfortunately, I feel some kind of apathy today and do not find it funny anymore. However, I do believe that others might find it entertaining nonetheless)

## STOP DOING `StringView path`

- `StringView` was not supposed to store paths
- 9 months after old string deprecation yet no good solution found for adding a zero terminator for syscalls on non-Serenity hosts
- Wanted to work with paths anyway? We have a tool for that: ByteBuffer
- "Yeah, we lost a filesystem race, so reading from Core::File returned EISDIR" - Statements dreamed up by the utterly Deranged

LOOK at what SerenityOS developers have been demanding your Respect for all this time with all the fancy libraries and full control over the stack
(This is real code, done by REAL SerenityOS developers)
```cpp
// NOTE: We have to ensure that the path is null-terminated.
DeprecatedString path_string = path;
int rc = ::openat(fd, path_string.characters(), options, mode);
```
?????
```cpp
auto canonicalized_path = TRY(String::from_deprecated_string(LexicalPath::canonicalized_path(path)));
TRY(tar_stream.add_link(canonicalized_path, statbuf.st_mode, TRY(Core::System::readlink(path))));
```
???????
```cpp
Core::DirIterator directory_iterator(directory_path, Core::DirIterator::Flags::SkipDots);
while (directory_iterator.has_next()) {
    auto name = directory_iterator.next_path();
    struct stat st = {};
    if (fstatat(directory_iterator.fd(), name.characters(), &st, AT_SYMLINK_NOFOLLOW) < 0)
        continue;
    bool is_directory = S_ISDIR(st.st_mode);
```
?????????????

"Hello I pretend that filesystem doesn't change while I am working with it"
They have played us for absolute fools

### CODE REVIEW

File: Userland/Libraries/LibCore/Forward.h [hunk 1]
```diff
@@ -10,15 +10,15 @@
 
 namespace Core {
 
-class AnonymousBuffer;
```
Comment (DanShaders) [inline Userland/Libraries/LibCore/Forward.h:13]: This is unintentional

**Comment by ADKaster** [comment]
Some thoughts:

- Ali seems to be on the path of converting DeprecatedfString --> CString or NullTerminatedString for those valid use cases
- My "3 years from now" perspective is that I'm not sure how we could adapt this to non-posix platforms (looking at you Win32)
   - but that doesn't mean it wouldn't be a good design for today!
- I was thinking that a Core::Path api would look more like what rust has: std::path::Path and std::path::PathBuf, where one is a slice, the other an owned buffer that points to our data. This class you've got seems more like a ... RelativePath?
- I guess I'm not clear on when users will want LexicalPath vs this Path.

Also I'm working towards a world where most applications that want to load resources (fonts, images, default files, etc) will load them via the new Core::Resource abstraction rather than directly via filesystem paths. That ofc doesn't apply to the use case you mentioned, of tar or other places where we actually do want to interact with the local filesystem (load from file, etc).

**Comment by DanShaders** [comment]
Not stale

**Comment by DanShaders** [comment]
Not stale

**Pull request closed without merging**